apps/req.c

```diff
@@ -1515,61 +1478,34 @@ static int join(char buf[], size_t buf_size, const char *name,
     return 1;
 }
 
-static EVP_PKEY_CTX *set_keygen_ctx(const char *gstr,
-                                    int *pkey_type, long *pkeylen,
-                                    char **palgnam, ENGINE *keygen_engine)
+static EVP_PKEY_CTX *set_keygen_ctx(const char *gstr, long default_keylen,
+                                    char **palgnam)
 {
     EVP_PKEY_CTX *gctx = NULL;
     EVP_PKEY *param = NULL;
-    long keylen = -1;
+    long keylen = default_keylen;
     BIO *pbio = NULL;
     const char *paramfile = NULL;
 
     if (gstr == NULL) {
-        *pkey_type = EVP_PKEY_RSA;
-        keylen = *pkeylen;
+        *palgnam = OPENSSL_strdup("RSA");
     } else if (gstr[0] >= '0' && gstr[0] <= '9') {
-        *pkey_type = EVP_PKEY_RSA;
+        *palgnam = OPENSSL_strdup("RSA");
         keylen = atol(gstr);
-        *pkeylen = keylen;
     } else if (strncmp(gstr, "param:", 6) == 0) {
         paramfile = gstr + 6;
     } else {
         const char *p = strchr(gstr, ':');
-        int len;
-        ENGINE *tmpeng;
-        const EVP_PKEY_ASN1_METHOD *ameth;
 
-        if (p != NULL)
-            len = p - gstr;
-        else
-            len = strlen(gstr);
-        /*
-         * The lookup of a the string will cover all engines so keep a note
-         * of the implementation.
-         */
-
-        ameth = EVP_PKEY_asn1_find_str(&tmpeng, gstr, len);
-
-        if (ameth == NULL) {
-            BIO_printf(bio_err, "Unknown algorithm %.*s\n", len, gstr);
-            return NULL;
-        }
+        if (p != NULL) {
+            *palgnam = OPENSSL_strndup(gstr, p - gstr);
 
-        EVP_PKEY_asn1_get0_info(NULL, pkey_type, NULL, NULL, NULL, ameth);
-#ifndef OPENSSL_NO_ENGINE
-        finish_engine(tmpeng);
-#endif
-        if (*pkey_type == EVP_PKEY_RSA) {
-            if (p != NULL) {
+            if (!strcasecmp(*palgnam, "RSA") || !strcasecmp(*palgnam, "RSA-PSS"))
                 keylen = atol(p + 1);
-                *pkeylen = keylen;
-            } else {
-                keylen = *pkeylen;
-            }
-        } else if (p != NULL) {
-            paramfile = p + 1;
-        }
+            else
+                paramfile = p + 1;
+        } else
```

Comment:
Nit: our coding style requires that if one branch of an `if` statement is wrapped with braces, then so must the other branch.